File: implementations/go/sorting/cocktail/cocktail_sort.go
package main

import (
  "fmt"
  "reflect"
)

func cocktailShakerSort(values []int) []int {
  a := append([]int(nil), values...)
  start, end := 0, len(a)-1
  swapped := true
  for swapped {
    swapped = false
    for i := start; i < end; i++ { if a[i] > a[i+1] { a[i], a[i+1] = a[i+1], a[i]; swapped = true } }
    if !swapped { break }
    swapped = false
    end--
    for i := end; i > start; i-- { if a[i-1] > a[i] { a[i-1], a[i] = a[i], a[i-1]; swapped = true } }
    start++
  }
  return a
}
func main() {
  if !reflect.DeepEqual(cocktailShakerSort([]int{5,1,4,2,8,0,-3}), []int{-3,0,1,2,4,5,8}) { panic("primary case failed") }
  if !reflect.DeepEqual(cocktailShakerSort([]int{3,3,2,1}), []int{1,2,3,3}) { panic("duplicate case failed") }
  fmt.Println("go cocktail ok")
}

File: implementations/fortran/sorting/cocktail/cocktail_sort.f90
program cocktail_sort_test
  implicit none
  integer :: a(7) = [5, 1, 4, 2, 8, 0, -3]
  integer :: e(7) = [-3, 0, 1, 2, 4, 5, 8]
  integer :: b(4) = [3, 3, 2, 1]
  integer :: f(4) = [1, 2, 3, 3]
  call cocktail_shaker_sort(a)
  if (any(a /= e)) error stop 'primary case failed'
  call cocktail_shaker_sort(b)
  if (any(b /= f)) error stop 'duplicate case failed'
  print *, 'fortran cocktail ok'
contains
  subroutine cocktail_shaker_sort(values)
    integer, intent(inout) :: values(:)
    integer :: start, finish, i, tmp
    logical :: swapped
    start = 1
    finish = size(values)
    swapped = .true.
    do while (swapped)
      swapped = .false.
      do i = start, finish - 1
        if (values(i) > values(i + 1)) then
          tmp = values(i); values(i) = values(i + 1); values(i + 1) = tmp
          swapped = .true.
        end if
      end do
      if (.not. swapped) exit
      swapped = .false.
      finish = finish - 1
      do i = finish, start + 1, -1
        if (values(i - 1) > values(i)) then
          tmp = values(i - 1); values(i - 1) = values(i); values(i) = tmp
          swapped = .true.
        end if
      end do
      start = start + 1
    end do
  end subroutine cocktail_shaker_sort
end program cocktail_sort_test
